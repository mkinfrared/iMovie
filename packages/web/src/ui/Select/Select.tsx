import FormControl from "@material-ui/core/FormControl";
import FormHelperText from "@material-ui/core/FormHelperText";
import InputLabel from "@material-ui/core/InputLabel";
import MenuItem from "@material-ui/core/MenuItem";
import MaterialSelect from "@material-ui/core/Select";
import React from "react";

import { SelectProps } from "./Select.type";

const Select = <T,>({
  className,
  ref,
  label,
  options,
  onChange,
  helperText,
  defaultValue,
  error = false
}: SelectProps<T>) => {
  const menuItems = options.map(({ label: name, value }) => (
    <MenuItem key={`${name}${value}`} value={value}>
      {name}
    </MenuItem>
  ));

  return (
    <FormControl
      className={className}
      variant="outlined"
      error={error}
      fullWidth
    >
      <InputLabel id="select">{label}</InputLabel>
      <MaterialSelect
        ref={ref}
        labelId="select"
        label={label}
        onChange={onChange as any}
        defaultValue={defaultValue}
      >
        {menuItems}
      </MaterialSelect>
      {helperText && <FormHelperText>{helperText}</FormHelperText>}
    </FormControl>
  );
};

export default React.memo(Select);
