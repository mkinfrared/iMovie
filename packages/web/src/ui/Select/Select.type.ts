export interface Option {
  label: string | number;
  value: string | number | string[];
}

export interface SelectProps<T> {
  label: string;
  options: Option[];
  onChange: (event: React.ChangeEvent<HTMLOptionElement>) => void;
  className?: string;
  ref?: React.Ref<T>;
  error?: boolean;
  helperText?: React.ReactNode;
}
