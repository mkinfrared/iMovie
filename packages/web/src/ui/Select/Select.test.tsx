import { render } from "@testing-library/react";
import React, { ReactElement } from "react";

import Select from "./Select";

describe("<Select />", () => {
  const className = "marklar";
  const options = [{ value: 42, label: "foobar" }];
  const helperText = "they took'er jobs!!";
  const label = "Cartman";
  const onChange = jest.fn();

  let Component: ReactElement;

  beforeEach(() => {
    Component = (
      <Select
        className={className}
        options={options}
        helperText={helperText}
        label={label}
        onChange={onChange}
      />
    );
  });

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("should match a snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });
});
