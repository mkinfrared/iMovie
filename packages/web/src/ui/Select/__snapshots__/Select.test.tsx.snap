// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should match a snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root marklar MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink="false"
      id="select"
    >
      Cartman
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="select"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        role="button"
        tabindex="0"
      >
        <span>
          â€‹
        </span>
      </div>
      <input
        type="hidden"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-7"
        >
          <span>
            Cartman
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-contained"
    >
      they took'er jobs!!
    </p>
  </div>
</div>
`;
